// app-theme/lib/components/MenuBar/const.js
export const menuCloseEvent = 'app-theme.components.MenuBar.Close';
export const menuElementMouseEnterEvent = 'app-theme.components.MenuBar.MenuElement.MouseEnter';
export const menuTopElementHorizontalEvent = 'app-theme.components.MenuBar.MenuTopElement.MoveHorizontal';
export const menuTopElementSwitchEvent = 'app-theme.components.MenuBar.MenuTopElement.SwitchElement';
export var Direction;
(function (Direction) {
    Direction["Down"] = "ArrowDown";
    Direction["Left"] = "ArrowLeft";
    Direction["Right"] = "ArrowRight";
    Direction["Up"] = "ArrowUp";
})(Direction || (Direction = {}));
export const onGenericKeyDown = (handleHorizontal, handleVertical) => (event) => {
    event.stopPropagation();
    const { currentTarget, key } = event;
    const allowedDirections = [Direction.Up, Direction.Down, Direction.Left, Direction.Right];
    if (!allowedDirections.includes(key)) {
        return;
    }
    const direction = key;
    if ((direction === Direction.Left || direction === Direction.Right)
        && handleHorizontal) {
        handleHorizontal(direction, currentTarget);
        return;
    }
    if ((direction === Direction.Up || direction === Direction.Down)
        && handleVertical) {
        handleVertical(direction, currentTarget);
    }
};
